docs/structured-streaming-programming-guide.md

```diff
@@ -1689,6 +1689,25 @@ hence the number is not same as the number of original input rows. You'd like to
 There's a known workaround: split your streaming query into multiple queries per stateful operator, and ensure
 end-to-end exactly once per query. Ensuring end-to-end exactly once for the last query is optional.
 
+### State Store and task locality
+
+The stateful operations store states for events in state stores of executors. State stores occupy resources such as memory and disk space to store the states.
+So it is more efficient to keep a state store provider running in the same executor across different streaming batches.
+Changing the location of a state store provider requires loading from checkpointed states from HDFS in the new executor.
```

Comment:
Mentioning HDFS sounds slightly incorrect - even in built-in state store (named HDFS) you can store the checkpoint to the external storages including object stores.

Probably `loading from checkpoint` would just work - if it doesn't seem to emphasize the overhead, it's OK to explicitly mention the overhead, like `the overhead of loading state from checkpoint depends on the external storage and the size of the state, which tends to hurt the latency of micro-batch run`.